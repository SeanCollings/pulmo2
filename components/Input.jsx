import React, { useState } from 'react';
import { StyleSheet, View, Text, Animated, TextInput } from 'react-native';

import { useTheme } from '../hooks/useTheme';

const Input = ({
  value,
  onChangeText,
  customStyles,
  maxLength = 40,
  showRemaining = false,
  ...rest
}) => {
  const theme = useTheme();
  const [animatedGrow] = useState(new Animated.Value(0));
  const [animationOpacity] = useState(new Animated.Value(0));

  const interpolateGrow = animatedGrow.interpolate({
    inputRange: [0, 1],
    outputRange: ['0%', '100%'],
  });

  const opacity = theme.DARK ? 0.87 : 1;
  const opacityDarker = theme.DARK ? 0.63 : 0.8;

  const onFocusHandler = () => {
    Animated.parallel([
      Animated.timing(animatedGrow, {
        toValue: 1,
        duration: 100,
      }),
      Animated.timing(animationOpacity, {
        toValue: opacity,
        duration: 100,
      }),
    ]).start();
  };
  const onBlurHandler = () => {
    Animated.parallel([
      Animated.timing(animatedGrow, {
        toValue: 0,
        duration: 10,
      }),
      Animated.timing(animationOpacity, {
        toValue: 0,
        duration: 10,
      }),
    ]).start();
  };

  const onChangeTextHandler = (text) => {
    let newText = text;
    if (newText > maxLength) {
      newText.substr(0, maxLength);
    }
    onChangeText(text);
  };

  const animatedGrowStyle = {
    width: interpolateGrow,
    opacity: animationOpacity,
  };

  return (
    <View style={styles.container}>
      <View style={styles.textInputContainer}>
        <TextInput
          {...rest}
          style={{
            ...styles.input,
            color: theme.TEXT,
            opacity,
            ...customStyles,
          }}
          value={value}
          onChangeText={onChangeTextHandler}
          maxLength={maxLength}
          numberOfLines={2}
          onFocus={onFocusHandler}
          onBlur={onBlurHandler}
          multiline
          selectionColor={theme.SECONDARY}
        />
        <Animated.View
          style={{
            borderBottomWidth: 1,
            borderBottomColor: theme.BORDER,
            ...animatedGrowStyle,
          }}
        ></Animated.View>
      </View>
      <View style={styles.remainingContainer}>
        <Text
          style={{
            ...styles.remainingText,
            color: theme.TEXT,
            opacity: opacityDarker,
          }}
        >
          {`${value.length} / ${maxLength}`}
        </Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
    paddingHorizontal: 5,
  },
  textInputContainer: {
    width: '78%',
  },
  input: {
    fontFamily: 'tit-light',
    fontSize: 17,
  },
  remainingContainer: {
    width: '22%',
    justifyContent: 'center',
    alignItems: 'center',
    paddingLeft: 10,
  },
  remainingText: {
    fontFamily: 'tit-light',
    fontSize: 14,
  },
});

export default Input;
