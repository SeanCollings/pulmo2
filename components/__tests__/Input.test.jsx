import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import Input from '../Input';

describe('Input - unit test', () => {
  let component;
  const value = 'Input';

  afterEach(() => {
    cleanup();
  });

  test('should render', async () => {
    const onChangeText = () => {};

    await act(async () => {
      component = create(<Input value={value} onChangeText={onChangeText} />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should handle text input', async () => {
    const onChangeText = jest.fn();

    await act(async () => {
      component = create(<Input value={value} onChangeText={onChangeText} />);
    });

    const textInputHandler = component.root.find(
      ({ props }) => props.testID === 'textInput'
    );

    fireEvent(textInputHandler, 'onChangeText', 'N');
    expect(onChangeText).toHaveBeenCalled();
  });

  test('should not allow text input of length greater than the maximum', async () => {
    const onChangeText = jest.fn();

    await act(async () => {
      component = create(
        <Input value={value} onChangeText={onChangeText} maxLength={2} />
      );
    });

    const textInputHandler = component.root.find(
      ({ props }) => props.testID === 'textInput'
    );

    fireEvent(textInputHandler, 'onChangeText', 'New Text');
    expect(onChangeText).toHaveBeenCalledWith('Ne');
  });
});
