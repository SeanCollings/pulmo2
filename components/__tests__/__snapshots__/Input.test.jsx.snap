// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input - unit test should render 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 5,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "width": "78%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      maxLength={40}
      multiline={true}
      numberOfLines={2}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      rejectResponderTermination={true}
      selectionColor="#002f56"
      style={
        Object {
          "color": "#3e4b57",
          "fontFamily": "tit-light",
          "fontSize": 17,
          "opacity": 1,
        }
      }
      testID="textInput"
      underlineColorAndroid="transparent"
      value="Input"
    />
    <View
      style={
        Object {
          "borderBottomColor": "#CCC",
          "borderBottomWidth": 1,
          "opacity": 0,
          "width": "0%",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingLeft": 10,
        "width": "22%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3e4b57",
          "fontFamily": "tit-light",
          "fontSize": 14,
          "opacity": 0.8,
        }
      }
    >
      5 / 40
    </Text>
  </View>
</View>
`;
